h2/src/main/org/h2/expression/Function.java

```diff
@@ -1493,7 +1497,46 @@ private Value getValueWithArgs(Session session, Expression[] args) {
             break;
         case EXTRACT: {
             int field = getDatePart(v0.getString());
-            result = ValueInt.get(DateTimeUtils.getDatePart(v1, field));
+            // Normal case when we don't retrieve the EPOCH time
+            if(field != EPOCH) {
+                result = ValueInt.get(DateTimeUtils.getDatePart(v1, field));	
+            }else {
+            	// Case where we retrieve the EPOCH time.
+            	// First we retrieve the dateValue and his time in nanoseconds.
+            	long[] a = DateTimeUtils.dateAndTimeFromValue(v1);
+                long dateValue = a[0];
+                long timeNanos = a[1];
+                // We compute the time in nanoseconds and the total number of days.
+        		BigDecimal timeNanosBigDecimal = new BigDecimal(timeNanos);
```

Comment:
Use 4 spaces for indentation. Do not use tabulation character, change your settings in Eclipse for this project. Check also lines below.